docs/settings.md
# Settings

Below are described all the settings supported by torrest. One can get/set them through the `/settings` endpoint.

| name                   | type    | default                              | description                                                                                                                                                                                                                               |
|------------------------|---------|--------------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| listen_port            | int     | 6889                                 | The bittorrent service listen port                                                                                                                                                                                                        |
| listen_interfaces      | string  | 0.0.0.0:listen_port,[::]:listen_port | A comma-separated list of (IP or device name, port) pairs. When set to `auto:<interface-name[:port]>` (for instance, `auto:wlan0` or `auto:wlan0:61000,wlan1`), Torrest automatically finds the addresses associated with such interfaces |
| outgoing_interfaces    | string  |                                      | A comma-separated list of IP addresses and interface names                                                                                                                                                                                |
| disable_dht            | boolean | false                                | Disable the dht node                                                                                                                                                                                                                      |
| disable_upnp           | boolean | false                                | Disable the UPnP service                                                                                                                                                                                                                  |
| disable_natpmp         | boolean | false                                | Disable the NAT-PMP service                                                                                                                                                                                                               |
| disable_lsd            | boolean | false                                | Disable the Local Service Discovery                                                                                                                                                                                                       |
| download_path          | string  | downloads                            | The download path                                                                                                                                                                                                                         |
| torrents_path          | string  | downloads/torrents                   | The torrents, magnets and fast-resume files download path                                                                                                                                                                                 |
| user_agent             | int     | 0                                    | The client identification to the tracker                                                                                                                                                                                                  |
| session_save           | int     | 30                                   | The time between save resume data calls                                                                                                                                                                                                   |
| tuned_storage          | boolean | false                                | Whether to use tuned storage settings                                                                                                                                                                                                     |
| check_available_space  | boolean | true                                 | Whether to check available space on torrent download                                                                                                                                                                                      |
| connections_limit      | int     | 0                                    | The connections limit (if 0, torrest chooses what to set)                                                                                                                                                                                 |
| limit_after_buffering  | boolean | false                                | Whether to only apply download/upload rate limits when not buffering                                                                                                                                                                      |
| max_download_rate      | int     | 0                                    | Max download rate, in bytes per second (0 means unlimited)                                                                                                                                                                                |
| max_upload_rate        | int     | 0                                    | Max upload rate, in bytes per second (0 means unlimited)                                                                                                                                                                                  |
| share_ratio_limit      | int     | 0 [200]                              | The share ratio limit (bytes up / bytes down)                                                                                                                                                                                             |
| seed_time_ratio_limit  | int     | 0 [700]                              | The seed time ratio limite (seconds as seed / seconds as downloader)                                                                                                                                                                      |
| seed_time_limit        | int     | 0 [24 * 60 * 60]                     | The limit on the time a torrent has been an active seed                                                                                                                                                                                   |
| active_downloads_limit | int     | 3                                    | How many downloading torrents the queuing mechanism allows                                                                                                                                                                                |
| active_seeds_limit     | int     | 5                                    | How many seeding torrents the queuing mechanism allows                                                                                                                                                                                    |
| active_checking_limit  | int     | 1                                    | The limit of simultaneous checking torrents                                                                                                                                                                                               |
| active_dht_limit       | int     | 88                                   | The max number of torrents to announce to the DHT                                                                                                                                                                                         |
| active_tracker_limit   | int     | 1600                                 | The max number of torrents to announce to their trackers                                                                                                                                                                                  |
| active_lsd_limit       | int     | 60                                   | The max number of torrents to announce to the local network over the local service discovery protocol                                                                                                                                     |
| active_limit           | int     | 500                                  | Hard limit on the number of active torrents                                                                                                                                                                                               |
| encryption_policy      | int     | 0                                    | The encryption policy                                                                                                                                                                                                                     |
| proxy.type             | int     |                                      | The proxy type                                                                                                                                                                                                                            |
| proxy.port             | int     |                                      | The proxy port                                                                                                                                                                                                                            |
| proxy.hostname         | string  |                                      | The proxy hostname                                                                                                                                                                                                                        |
| proxy.username         | string  |                                      | The proxy username                                                                                                                                                                                                                        |
| proxy.passwrod         | string  |                                      | The proxy password                                                                                                                                                                                                                        |
| buffer_size            | int     | 20 * 1024 * 1024                     | The buffer size to consider when prioritizing pieces                                                                                                                                                                                      |
| piece_wait_timeout     | int     | 60                                   | The piece wait timeout (when serving files)                                                                                                                                                                                               |
| piece_expiration       | int     | 5                                    | How much time to keep an unused piece in memory (unused on legacy read piece)                                                                                                                                                             |
| service_log_level      | int     | 2                                    | The service log level                                                                                                                                                                                                                     |
| alerts_log_level       | int     | 5                                    | Alerts log level                                                                                                                                                                                                                          |
| api_log_level          | int     | 4                                    | The API log level                                                                                                                                                                                                                         |

## User agent options

| value | user-agent                                                |
|-------|-----------------------------------------------------------|
| 0     | torrest/<torrest-version> libtorrent/<libtorrent-version> |
| 1     | libtorrent/<libtorrent-version>                           |
| 2     | libtorrent (Rasterbar) 1.1.0                              |
| 3     | BitTorrent 7.5.0                                          |
| 4     | BitTorrent 7.4.3                                          |
| 5     | µTorrent 3.4.9                                            |
| 6     | µTorrent 3.2.0                                            |
| 7     | µTorrent 2.2.1                                            |
| 8     | Transmission 2.92                                         |
| 9     | Deluge 1.3.6.0                                            |
| 10    | Deluge 1.3.12.0                                           |
| 11    | Vuze 5.7.3.0                                              |

## Encryption policy

| value | encryption | description                                                                  |
|-------|------------|------------------------------------------------------------------------------|
| 0     | Enabled    | Encrypted connections are enabled, but non-encrypted connections are allowed |
| 1     | Disabled   | Only non-encrypted connections are allowed                                   |
| 2     | Forced     | Only encrypted connections are allowed                                       |

## Proxy type

| value | type                 |
|-------|----------------------|
| 0     | none                 |
| 1     | SOCKS4               |
| 2     | SOCKS5               |
| 3     | SOCKS5 with password |
| 4     | HTTP                 |
| 5     | HTTP with password   |
| 6     | i2p SAM proxy        |

## Log level

| value | level    |
|-------|----------|
| 0     | trace    |
| 1     | debug    |
| 2     | info     |
| 3     | warning  |
| 4     | error    |
| 5     | critical |
| 6     | off      |

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Fixed

- Fix listen_interfaces port matching

### Added

- Add support to `write_mode` setting. It supports `0: auto`, `1: pwrite` and `2: mmap_write` values, defaulting
  to `0: auto`.
- Automatically find interfaces addresses when `listen_interfaces` is set to `auto:<interface-name[:port]>` (for
  instance, `auto:wlan0` or `auto:wlan0:61000,wlan1`). Multiple values are supported and must be comma separated.

## [v0.0.5] - 03/01/2024

### Fixed

- Always include "Accept-Ranges" header in serve responses.
- Update to libtorrent 1.2.19

### Changed

- Make info hash case-insensitive

## [v0.0.4] - 30/04/2023

### Changed

- Allow setting a custom user-agent
- Downgrade to last libtorrent v1.2 (1.2.18)

## [v0.0.3] - 08/04/2023

### Added

- Allow building different major versions of libtorrent.
- New safer read piece mechanism (around piece alerts). Legacy read piece is still supported for libtorrent v1.
- Internal piece cache with a configurable piece expiration (defaults to 5 seconds).
- Allow building as a shared library. The below methods are now available:
    - int start_with_env()
    - int start(uint16_t port, String settings_path, int global_log_level)
    - void stop()
    - void clear_logging_sinks()
    - void add_logging_stdout_sink()
    - void add_logging_file_sink(String file_path, bool truncate)
    - void add_logging_callback_sink(log_callback_fn callback)
- C and Python bindings to the shared library.
- Environment variables support.
- New options to command line invocation: log pattern and log path.

### Changed

- Update libtorrent to v2.0.8.
- Update nlohmann json to v3.11.0.
- Improve piece wait timeout so it is always respected.
- Updated all API verbs for correctness.
- Only set max single core connections if connections_limit is not explicitly set (arm devices only).
- Improved proxy settings validation.

### Fixed

- Correctly close HTTP connections on server shutdown.
- Fix query parameters default values.

## [v0.0.2] - 05/06/2022

Hotfix release.

### Fixed

- Fix server error DTOs.
- Do not handle duplicate torrents as error when loading torrents.
- Disable atomic linking on oatpp for android and linux builds.

## [v0.0.1] - 29/05/2022

First release.

### Added

- Libtorrent 1.2.16 support.
- Multi-platform support with cross build environments: android-arm, android-arm64, android-x64, android-x86,
  darwin-x64, linux-armv7, linux-arm64, linux-x64, linux-x86, windows-x64 and windows-x86.
- REST API with swagger containing settings, service, torrents and files management endpoints.
- Configurable service with a comprehensive list of settings.
- Buffering functionality to prioritize certain pieces making them available first.

[Unreleased]: https://github.com/i96751414/torrest-cpp/compare/v0.0.5...master

[v0.0.5]: https://github.com/i96751414/torrest-cpp/compare/v0.0.4...v0.0.5

[v0.0.4]: https://github.com/i96751414/torrest-cpp/compare/v0.0.3...v0.0.4

[v0.0.3]: https://github.com/i96751414/torrest-cpp/compare/v0.0.2...v0.0.3

[v0.0.2]: https://github.com/i96751414/torrest-cpp/compare/v0.0.1...v0.0.2

[v0.0.1]: https://github.com/i96751414/torrest-cpp/commits/v0.0.1

src/settings/settings.h
#ifndef TORREST_SETTINGS_H
#define TORREST_SETTINGS_H

#include "libtorrent/aux_/export.hpp"
#include "spdlog/spdlog.h"

namespace torrest { namespace settings {

#if TORRENT_ABI_VERSION > 2
    enum WriteMode {
        wm_auto,
        wm_pwrite,
        wm_mmap_write,
        wm_num_values
    };
#endif

    enum EncryptionPolicy {
        ep_enabled,
        ep_disabled,
        ep_forced,
        ep_num_values
    };

    enum ProxyType {
        pt_none,
        pt_socks4,
        pt_socks5,
        pt_socks5_password,
        pt_http,
        pt_http_password,
        pt_i2psam,
        pt_num_values
    };

    struct ProxySettings {
        ProxyType type = pt_none;
        int port = 0;
        std::string hostname;
        std::string username;
        std::string password;

        void validate() const;
    };

    struct Settings {
        int listen_port = 6889;
        std::string listen_interfaces;
        std::string outgoing_interfaces;
        bool disable_dht = false;
        bool disable_upnp = false;
        bool disable_natpmp = false;
        bool disable_lsd = false;
        std::string download_path = "downloads";
        std::string torrents_path = "downloads/torrents";
        std::string user_agent;
        int session_save = 30;
        bool tuned_storage = false;
        bool check_available_space = true;
        int connections_limit = 0;
        bool limit_after_buffering = false;
        int max_download_rate = 0;
        int max_upload_rate = 0;
        int share_ratio_limit = 0;
        int seed_time_ratio_limit = 0;
        int seed_time_limit = 0;
        int active_downloads_limit = 3;
        int active_seeds_limit = 5;
        int active_checking_limit = 1;
        int active_dht_limit = 88;
        int active_tracker_limit = 1600;
        int active_lsd_limit = 60;
        int active_limit = 500;
#if TORRENT_ABI_VERSION > 2
        WriteMode write_mode = wm_auto;
#endif
        EncryptionPolicy encryption_policy = ep_enabled;
        std::shared_ptr<ProxySettings> proxy = nullptr;
        std::int64_t buffer_size = 20 * 1024 * 1024;
        int piece_wait_timeout = 60;
#if !TORREST_LEGACY_READ_PIECE
        int piece_expiration = 5;
#endif
        spdlog::level::level_enum service_log_level = spdlog::level::info;
        spdlog::level::level_enum alerts_log_level = spdlog::level::critical;
        spdlog::level::level_enum api_log_level = spdlog::level::err;

        static Settings load(const std::string &pPath);

        static Settings parse(const std::string &pJson);

        void save(const std::string &pPath) const;

        std::string dump() const;

        void validate() const;
    };

}}

#endif //TORREST_SETTINGS_H
